File: lintcode/palindrome-partitioning-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/palindrome-partitioning-ii/

class Solution:
    # @param s, a string
    # @return an integer
    def minCut(self, s):
        # write your code here
        if len(s) <= 0:
            return 0

        dp = [i for i in xrange(-1, len(s))]
        for i in xrange(len(s)):
            j = 0
            while i-j>=0 and i+j<len(s) and s[i-j] == s[i+j]:
                dp[i+j+1] = min(dp[i-j]+1, dp[i+j+1])
                j += 1

            j = 0
            while i-j>=0 and i+j+1<len(s) and s[i-j] == s[i+j+1]:
                dp[i+j+2] = min(dp[i-j]+1, dp[i+j+2])
                j += 1

        return dp[len(s)]

File: leetcode-2/palindrome-partitioning-ii.cc
// Question: https://leetcode.com/problems/palindrome-partitioning-ii/

class Solution {
public:
    int minCut(string s) {
        int len = s.size();
        if (len <= 1) {
            return 0;
        }

        vector<bool> is_pal(len, false);
        vector<int> dp(len+1, -1);
        for (int i = 1; i <= len; ++i) {
            dp[i] = dp[i-1] + 1;
        }

        for (int i = 1; i < len; ++i) {
            for (int j = 0; j <= i; ++j) {
                if (j == i) {
                    is_pal[j] = true;
                } else if (j == i-1) {
                    is_pal[j] = s[j] == s[i];
                } else {
                    is_pal[j] = (s[j] == s[i] && is_pal[j+1]);
                }
                if (is_pal[j]) {
                    dp[i+1] = min(dp[i+1], dp[j]+1);
                }
            }
        }
        return dp.back();
    }
};
